File: C++/space/triangulation_view_performance.cpp

#include <cmath>
#include <map>
#include <random>
#include <set>

#include "datastructures/multi_tree_view.hpp"
#include "space/initial_triangulation.hpp"
#include "space/triangulation.hpp"
#include "space/triangulation_view.hpp"

int bsd_rnd() {
  static unsigned int seed = 0;
  int a = 1103515245;
  int c = 12345;
  unsigned int m = 2147483648;
  return (seed = (a * seed + c) % m);
}

using namespace space;
using namespace datastructures;

constexpr int level = 10;
constexpr int iters = 300;

int main() {
  std::cout << "Sizeof comparison" << std::endl;
  std::cout << "unsigned int : " << sizeof(unsigned int) << std::endl;
  std::cout << "int : " << sizeof(int) << std::endl;
  std::cout << "uint : " << sizeof(uint) << std::endl;
  std::cout << "size_t : " << sizeof(size_t) << std::endl;
  std::cout << "Vertex * : " << sizeof(Vertex *) << std::endl;
  std::cout << "std::vector<Vertex *> : " << sizeof(std::vector<Vertex *>)
            << std::endl;
  std::cout << "SmallVector<Vertex *,4> : " << sizeof(SmallVector<Vertex *, 4>)
            << std::endl;
  std::cout << "SmallVector<Vertex *,8> : " << sizeof(SmallVector<Vertex *, 8>)
            << std::endl;
  std::cout << "StaticVector<Vertex *,4> : "
            << sizeof(StaticVector<Vertex *, 4>) << std::endl;
  std::cout << "std::array<Vertex *, 4> : " << sizeof(std::array<Vertex *, 4>)
            << std::endl;
  auto T = InitialTriangulation::UnitSquare();
  T.elem_tree.UniformRefine(::level);

  for (size_t i = 0; i < iters; ++i) {
    // Create a random subtree
    auto vertex_subtree = TreeView<Vertex>(T.vertex_meta_root);
    vertex_subtree.DeepRefine(
        /* call_filter */ [](auto &&vertex) {
          return vertex->level() <= 0 || bsd_rnd() % 3 != 0;
        });

    auto T_view = TriangulationView(vertex_subtree);
  }
}

File: Python/space/triangulation_view_performance.py
from ..datastructures.tree_view import TreeView
from .triangulation import InitialTriangulation
from .triangulation_view import TriangulationView


def bsd_rand(seed):
    def rand():
        nonlocal seed
        seed = (1103515245 * seed + 12345) & 0x7fffffff
        return seed

    return rand


def test_python(level, iters):
    bsd_rnd = bsd_rand(0)
    T = InitialTriangulation.unit_square()
    T.elem_meta_root.uniform_refine(level)

    for _ in range(iters):
        vertex_subtree = TreeView.from_metaroot(T.vertex_meta_root)
        vertex_subtree.deep_refine(call_filter=lambda vertex: vertex.level <= 0
                                   or (bsd_rnd() % 3) != 0)
        T_view = TriangulationView(vertex_subtree)


def test_cppyy(level, iters):
    import cppyy
    cppyy.cppdef('#pragma cling optimize 3')
    cppyy.include('C++/space/triangulation_view.hpp')
    cppyy.cppdef('extern template datastructures::NodeView<space::Vertex>;')
    cppyy.load_library('C++/build/space/libtriangulation.dylib')
    cppyy.load_library('C++/build/space/libtriangulation_view_lib.dylib')
    bsd_rnd = bsd_rand(0)
    T = cppyy.gbl.space.InitialTriangulation.UnitSquare()
    T.elem_meta_root.UniformRefine(level)

    for _ in range(iters):
        vertex_subtree = cppyy.gbl.datastructures.NodeView[
            cppyy.gbl.space.Vertex].CreateRoot(T.vertex_meta_root)
        vertex_subtree.DeepRefine()
        T_view = cppyy.gbl.space.TriangulationView(vertex_subtree)


if __name__ == "__main__":
    test_python(level=10, iters=500)
    #test_cppyy(level=10, iters=10000)
